### PR TITLE
Support serialization of  `astropy.wcs.WCS` objects to ASDF. 

### PR DESCRIPTION
**Pull request by ViciousEagle03**
This PR is a [WIP] and adds the support for serialization of `astropy.wcs.WCS` to ASDF. 
fixes: #234 
Checklist: 

- [x] Add support for the basic WCSes by preserving the `wcs.to_fits()`. 
- [x] Test

### CODE REVIEW

**Comment by braingram** [comment]
FYI I had to "approve and run" the CI for this (likely because this is your first contribution). Let me know if the CI stops running and I can look into fixing it. Thanks!

**Comment by ViciousEagle03** [comment]
Thanks for approving the CI run.

**Comment by ViciousEagle03** [comment]
The CI still needs to be approved to run.

**Comment by braingram** [comment]
Hmmm, I don't know of a way to "approve all runs for this PR from a first time contributor" and this suggests it might not be an option:
https://github.com/orgs/community/discussions/14334

Would it be possible to open a PR with a pretty minimal change that I can approve (which will then make you a contributor)? A PR that just removes the if check here:
https://github.com/astropy/asdf-astropy/blob/9348f899f57114b574b050ea59d3c74b28e498f6/asdf_astropy/converters/transform/tests/test_transform.py#L81
and unindents the code in the if block would work. The minimum required version of astropy is now 5.2 so the if will always be `True`:
https://github.com/astropy/asdf-astropy/blob/9348f899f57114b574b050ea59d3c74b28e498f6/pyproject.toml#L24

**Comment by ViciousEagle03** [comment]
Sure I am on it

**Comment by ViciousEagle03** [comment]
pre-commit.ci autofix

**Comment by ViciousEagle03** [comment]
Hey @braingram, to enable serialization for `SlicedLowLevelWCS`, we need the schema and converter for `fitswcs`. Should I push the serialization logic for `SlicedLowLevelWCS` here, or should I add the changes into a new branch after this is merged?

**Comment by braingram** [comment]
I'm ok having both of those features in this PR since they seem very related.

File: asdf_astropy/converters/fits/fitswcs.py
Comment (braingram) [inline asdf_astropy/converters/fits/fitswcs.py:9]:
```suggestion
    by preserving the ``wcs.to_header()`` data. It does not support complex WCS objects
```
This should hopefully fix the docs.

File: asdf_astropy/resources/manifests/astropy-1.0.0.yaml
Comment (braingram) [inline asdf_astropy/resources/manifests/astropy-1.0.0.yaml:31]: Would you add this to a new manifest `astropy-1.1.0.yaml` and register it as a new extension?

File: asdf_astropy/resources/schemas/fits/fitswcs-1.0.0.yaml
Comment (braingram) [inline asdf_astropy/resources/schemas/fits/fitswcs-1.0.0.yaml:17]: Can this be made more specific? Having a generic `object` will pass just about anything.

File: asdf_astropy/resources/schemas/fits/fitswcs-1.0.0.yaml
Comment (braingram) [inline asdf_astropy/resources/schemas/fits/fitswcs-1.0.0.yaml:7]:
```suggestion
title: Represents the fits object
```

Would you add a more descriptive title?

File: asdf_astropy/resources/schemas/fits/fitswcs-1.0.0.yaml
Comment (braingram) [inline asdf_astropy/resources/schemas/fits/fitswcs-1.0.0.yaml:10]:
```suggestion
description: >-
  Represents the fits object
```
Same as `title` would you add a more descriptive description?

File: asdf_astropy/resources/manifests/astropy-1.0.0.yaml
Comment (ViciousEagle03) [reply asdf_astropy/resources/manifests/astropy-1.0.0.yaml:31]: Sure

File: asdf_astropy/converters/slicedwcs/slicedwcs.py
Comment (braingram) [inline asdf_astropy/converters/slicedwcs/slicedwcs.py:32]: Would you make these slices dicts of `{'start': start, 'stop': stop, 'step': step}` to avoid the `literal_eval` calls and allow them to be defined in the schema?

File: asdf_astropy/converters/fits/fitswcs.py
Comment (braingram) [inline asdf_astropy/converters/fits/fitswcs.py:29]: This loses the header comments. I'm not opposed to that (for now) but is there an easy way to preserve these?

**Comment by braingram** [comment]
Can this use `WCS.to_fits` instead of `to_header` and use the existing [FitsConverter](https://github.com/astropy/asdf-astropy/blob/fa5f1f452412fc969b79ead4a0471d38dfa3a9cc/asdf_astropy/converters/fits/fits.py#L22)?

File: asdf_astropy/resources/schemas/fits/fitswcs-1.0.0.yaml
Comment (ViciousEagle03) [reply asdf_astropy/resources/schemas/fits/fitswcs-1.0.0.yaml:13]: I thought specifying the `tag` was a common convention in `asdf-astropy` schemas

Comment (braingram) [reply asdf_astropy/resources/schemas/fits/fitswcs-1.0.0.yaml:13]: If not needed (which I think is the case for the fits schema) it's more flexible to leave the `tag` out of the schema. I opened an issue for the fits schema https://github.com/astropy/asdf-astropy/issues/240 Please let me know if you see any other schemas that define a top-level `tag`.

**Comment by ViciousEagle03** [comment]
> Can this use `WCS.to_fits` instead of `to_header` and use the existing [FitsConverter](https://github.com/astropy/asdf-astropy/blob/fa5f1f452412fc969b79ead4a0471d38dfa3a9cc/asdf_astropy/converters/fits/fits.py#L22)?

Stuart and I had discussed using to_header() for the simpler WCS cases.

**Comment by Cadair** [comment]
I did not know the FITSConverter existed, so maybe that is easier. I shall have a look in more detail tomorrow.

**Comment by Cadair** [comment]
Thinking about this a bit more +1 to just using `to_fits`, it's a much better solution.

File: asdf_astropy/converters/fits/fitswcs.py
Comment (Cadair) [reply asdf_astropy/converters/fits/fitswcs.py:29]: Does the WCS object store the comments at all? I didn't think it did.

**Comment by Cadair** [comment]
> Hmmm, I don't know of a way to "approve all runs for this PR from a first time contributor" and this suggests it might not be an option:

FTR, the best solution is to change the settings in the repo settings to only require approval for new github accounts.

File: asdf_astropy/converters/slicedwcs/slicedwcs.py
Comment (Cadair) [reply asdf_astropy/converters/slicedwcs/slicedwcs.py:32]: You could go one further and write a converter for the built in `slice` object if you really wanted :stuck_out_tongue:

File: asdf_astropy/resources/manifests/astropy-1.0.0.yaml
Comment (ViciousEagle03) [reply asdf_astropy/resources/manifests/astropy-1.0.0.yaml:31]: `astropy-1.1.0.yaml` manifest already exists. Do I need to add it to this manifest?

File: asdf_astropy/converters/fits/tests/test_fitswcs.py
Comment (ViciousEagle03) [reply asdf_astropy/converters/fits/tests/test_fitswcs.py:10]: I have extracted the `WCS` from the `FITS` file. I thought it would be better to test directly using a `FITS` file.

File: asdf_astropy/converters/fits/fitswcs.py
Comment (ViciousEagle03) [reply asdf_astropy/converters/fits/fitswcs.py]: Since we are now using `to_fits` instead of `to_header`, we no longer need to specify that the converter supports only certain types of `WCSes`.

**Comment by ViciousEagle03** [comment]
Hello, @braingram could you please review the recent changes.

File: asdf_astropy/resources/manifests/astropy-1.0.0.yaml
Comment (braingram) [reply asdf_astropy/resources/manifests/astropy-1.0.0.yaml]: Yeah a new manifest is needed (I forgot this already has an 1.1.0).

Things are a bit more complicated as 1.2.0 exists but is only ASDF standard 1.6.0 compliant.

Would you add a new one (1.3.0) based off of the 1.1.0 one (which uses the current ASDF standard 1.5.0). I'll sort out the 1.6.0 issues after this PR is in.

File: asdf_astropy/converters/fits/fitswcs.py
Comment (braingram) [reply asdf_astropy/converters/fits/fitswcs.py]: Yes the comments are preserved. Reading an `astropy.wcs.WCS` from a random JWST file and calling `wcs.to_header` I see the comments.
```
WCSAXES =                    2 / Number of coordinate axes
CRPIX1  =                160.5 / Pixel coordinate of reference point
CRPIX2  =                160.5 / Pixel coordinate of reference point
...
```

File: asdf_astropy/converters/fits/fitswcs.py
Comment (braingram) [inline asdf_astropy/converters/fits/fitswcs.py:11]:
```suggestion
        primary_hdu = node["hdu"][0]
```

Doesn't this throw out the distortion (and other non-primary) data that is generated with `to_fits`? Would you add a test with a wcs that generates more than 1 hdu for `to_fits`?

File: asdf_astropy/converters/fits/tests/test_fitswcs.py
Comment (braingram) [inline asdf_astropy/converters/fits/tests/test_fitswcs.py:12]: This would introduce downloading data files as part of the test runs. This is not something done by other tests and I don't think this belongs in unit tests.

Would you add a fixture that creates suitable wcs objects?

File: asdf_astropy/converters/slicedwcs/slicedwcs.py
Comment (braingram) [inline asdf_astropy/converters/slicedwcs/slicedwcs.py:24]: When is this not a slice?

File: asdf_astropy/resources/manifests/astropy-1.1.0.yaml
Comment (braingram) [inline asdf_astropy/resources/manifests/astropy-1.1.0.yaml:60]: Would you make the schema title/descriptions match the ones here?

**Comment by braingram** [comment]
I left a few comments. Any clue why oldest deps is failing?

File: asdf_astropy/converters/slicedwcs/slicedwcs.py
Comment (ViciousEagle03) [reply asdf_astropy/converters/slicedwcs/slicedwcs.py:24]: If we examine the first test case from [test_slicedwcs.py](https://github.com/astropy/asdf-astropy/pull/235/files#:~:text=wcs0%20%3D%20SlicedLowLevelWCS(wcs%2C%201)), the initial element of the `slices_array` will be an integer:
```
(Pdb) wcs0 = SlicedLowLevelWCS(wcs, 1)
(Pdb) wcs0._slices_array
[1, slice(None, None, None), slice(None, None, None), slice(None, None, None)]
(Pdb)
```

File: asdf_astropy/resources/manifests/astropy-1.1.0.yaml
Comment (ViciousEagle03) [reply asdf_astropy/resources/manifests/astropy-1.1.0.yaml:60]: Sure

File: asdf_astropy/converters/fits/fitswcs.py
Comment (ViciousEagle03) [reply asdf_astropy/converters/fits/fitswcs.py:11]: @braingram, Do you have any suggestions for effectively preserving distortion data?

**Comment by ViciousEagle03** [comment]
> I left a few comments. Any clue why oldest deps is failing?

I suspect the [first error](https://github.com/astropy/asdf-astropy/actions/runs/10546063365/job/29217159635?pr=235#:~:text=FAILED%20asdf_astropy/converters/fits/tests/test_fitswcs.py%3A%3Atest_wcs_serialization%5Bwcs1%5D%20%2D%20ValueError%3A%20ERROR%205%20in%20wcsset()%20at%20line%202775%20of%20file%20cextern/wcslib/C/wcs.c%3A) is most likely due to the distortion data not being properly handled by the `fitswcs` converter.

File: asdf_astropy/converters/slicedwcs/slicedwcs.py
Comment (braingram) [reply asdf_astropy/converters/slicedwcs/slicedwcs.py:24]: Is there a public API for accessing the slices?

Comment (ViciousEagle03) [reply asdf_astropy/converters/slicedwcs/slicedwcs.py:24]: I don't think there is a public API for accessing the slices.

File: asdf_astropy/converters/fits/fitswcs.py
Comment (ViciousEagle03) [reply asdf_astropy/converters/fits/fitswcs.py:11]: I am currently experimenting and trying to determine the appropriate serialization logic for this [WCS](https://github.com/astropy/astropy/blob/d31407ddb99904263448d13bf87659b79d1cc0cc/astropy/wcs/tests/test_wcs.py#L1875C1-L1875C34). If you have any suggestions, please let me know.

File: asdf_astropy/converters/fits/tests/test_fitswcs.py
Comment (ViciousEagle03) [inline asdf_astropy/converters/fits/tests/test_fitswcs.py:7]: Should I create a separate fixture for the WCS test cases, or will the existing ones work?

File: asdf_astropy/converters/fits/tests/test_fitswcs.py
Comment (ViciousEagle03) [reply asdf_astropy/converters/fits/tests/test_fitswcs.py]: Please see [this](https://github.com/astropy/asdf-astropy/pull/235#discussion_r1734940367)

File: asdf_astropy/converters/fits/tests/test_fitswcs.py
Comment (braingram) [inline asdf_astropy/converters/fits/tests/test_fitswcs.py:68]:
```suggestion
        assert wcs.to_header() == loaded_wcs.to_header()
```

This isn't testing the distortion information for the wcses.
I tried changing this to `wcs == loaded_wcs` and this is failing in part due to the sip attribute not round-tripping.

File: asdf_astropy/converters/fits/fitswcs.py
Comment (braingram) [inline asdf_astropy/converters/fits/fitswcs.py:16]:
```suggestion
            node["hdu"] = wcs.to_fits(relax=True)
```
Presumably the relax is here for the "non-standard" wcs keywords. Will these only be present if `sip` is not None?

File: asdf_astropy/converters/slicedwcs/test/test_slicedwcs.py
Comment (braingram) [inline asdf_astropy/converters/slicedwcs/test/test_slicedwcs.py:25]: What's triggering this warning?

File: asdf_astropy/resources/manifests/astropy-1.3.0.yaml
Comment (braingram) [inline asdf_astropy/resources/manifests/astropy-1.3.0.yaml:9]: Why restrict the standard here?

File: asdf_astropy/resources/schemas/slicedwcs/slicedwcs-1.0.0.yaml
Comment (braingram) [inline asdf_astropy/resources/schemas/slicedwcs/slicedwcs-1.0.0.yaml:24]:
```suggestion
            - type: object
```
This needs more details. It appears to be an object with `start` `stop` `step` all either int or None.

File: asdf_astropy/converters/slicedwcs/slicedwcs.py
Comment (braingram) [inline asdf_astropy/converters/slicedwcs/slicedwcs.py:1]: I believe the general layout for this package is to put converters in submodules similar to the layout in astropy. Would you move this and `fitswcs.py` (and the corresponding tests) into a `wcs` submodule? Having them both contained in the new `wcs` submodule is sufficient (so the `SlicedWCSConverter` doesn't have to be further nested in a `wcsapi.wrappers.sliced_wcs` submodule).

File: asdf_astropy/converters/fits/tests/test_fitswcs.py
Comment (braingram) [reply asdf_astropy/converters/fits/tests/test_fitswcs.py:68]: The `wcs == loaded_wcs` appears to be defaulting to `object.__eq__` (so will fail since `id(wcs) != id(loaded_wcs)`). This helper function could be used instead:
https://github.com/astropy/asdf-astropy/blob/fa5f1f452412fc969b79ead4a0471d38dfa3a9cc/asdf_astropy/testing/helpers.py#L105
and reports a failure for this test since the cards don't match:
```
-> assert tuple(card_a) == tuple(card_b)
(Pdb) p card_a
('A_0_0', -2.0413459666638e-06, 'SIP distortion coefficient')
(Pdb) p card_b
('A_0_0', -2.0413459666638133e-06, 'SIP distortion coefficient')
```

File: asdf_astropy/converters/fits/tests/test_fitswcs.py
Comment (braingram) [inline asdf_astropy/converters/fits/tests/test_fitswcs.py:8]:
```suggestion
    rng = np.random.default_rng(42)
```
To make the tests deterministic.

**Review by braingram — CHANGES_REQUESTED** [review]

File: asdf_astropy/converters/slicedwcs/slicedwcs.py
Comment (ViciousEagle03) [reply asdf_astropy/converters/slicedwcs/slicedwcs.py:29]: Sure

File: asdf_astropy/converters/slicedwcs/test/test_slicedwcs.py
Comment (ViciousEagle03) [reply asdf_astropy/converters/slicedwcs/test/test_slicedwcs.py:25]: Same as the above reason

File: asdf_astropy/resources/manifests/astropy-1.3.0.yaml
Comment (ViciousEagle03) [reply asdf_astropy/resources/manifests/astropy-1.3.0.yaml:9]: Let me correct it

File: asdf_astropy/resources/schemas/slicedwcs/slicedwcs-1.0.0.yaml
Comment (ViciousEagle03) [reply asdf_astropy/resources/schemas/slicedwcs/slicedwcs-1.0.0.yaml:24]: Sure, I will include it in the schema

File: asdf_astropy/converters/slicedwcs/slicedwcs.py
Comment (ViciousEagle03) [reply asdf_astropy/converters/slicedwcs/slicedwcs.py:1]: Agreed, moving them to a `wcs` submodule makes sense and will keep things organized.

File: asdf_astropy/converters/fits/fitswcs.py
Comment (ViciousEagle03) [reply asdf_astropy/converters/fits/fitswcs.py:16]: Is there a case we are not considering? If so, could you please explain how we can account for it?

File: asdf_astropy/converters/fits/tests/test_fitswcs.py
Comment (ViciousEagle03) [reply asdf_astropy/converters/fits/tests/test_fitswcs.py:68]: Since there isn't a helper function available to check if two WCS objects are equivalent, I thought of checking the round-tripping of the WCS through its `header`.

**Comment by ViciousEagle03** [comment]
@braingram, @Cadair I'm sorry for being unavailable for the past two weeks because of my exams and practicals. Now that they are completed, I can fully dedicate my time to completing the project.

File: asdf_astropy/resources/schemas/slicedwcs/slicedwcs-1.0.0.yaml
Comment (braingram) [reply asdf_astropy/resources/schemas/slicedwcs/slicedwcs-1.0.0.yaml:24]: Thanks! The updated schema looks great.

**Comment by nden** [comment]
For the tests in this PR, would it be useful to use some of the data in the astropy.wcs.tests.data directory. There is a variety of WCSs there, including some with [non-sip distortion](https://github.com/astropy/astropy/blob/main/astropy/wcs/tests/data/dist_lookup.fits.gz).

Also would be be helpful to add a `description` to the fields in the schemas? And title?

**Comment by braingram** [comment]
Closing in favor of https://github.com/astropy/asdf-astropy/pull/246

File: asdf_astropy/converters/wcs/fitswcs.py
Comment (braingram) [reply asdf_astropy/converters/wcs/fitswcs.py:18]: I don't think so. https://github.com/astropy/asdf-astropy/pull/246/files#diff-ff1eab3e17c709f13beb01b87af0c1b0c8d6cc332436c4b7c3564673ecda950aR37 has that suggestion.

File: asdf_astropy/converters/wcs/fitswcs.py
Comment (braingram) [reply asdf_astropy/converters/wcs/fitswcs.py:15]: Are you referring to the "table" and "TAB" distortions or is there another type?
- "TAB" is a non-starter since astropy can't write them: https://github.com/astropy/astropy/issues/9998
- "table" is tested: https://github.com/astropy/asdf-astropy/pull/246/files#diff-4d12199a58487bbc9c9a729bcb6f8a3d6297a6547eb1510c29c6a8721986b43eR40

**Pull request closed without merging**
